src/wallet/wallet.cpp

```diff
@@ -4124,3 +4126,88 @@ bool CWalletTx::AcceptToMemoryPool(const CAmount& nAbsurdFee, CValidationState&
     fInMempool = ret;
     return ret;
 }
+
+static const std::string OUTPUT_TYPE_STRING_LEGACY = "legacy";
+static const std::string OUTPUT_TYPE_STRING_P2SH = "p2sh";
+static const std::string OUTPUT_TYPE_STRING_BECH32 = "bech32";
+
+OutputType ParseOutputType(const std::string& type, OutputType default_type)
+{
+    if (type == "") {
+        return default_type;
+    } else if (type == OUTPUT_TYPE_STRING_LEGACY) {
+        return OUTPUT_TYPE_LEGACY;
+    } else if (type == OUTPUT_TYPE_STRING_P2SH) {
+        return OUTPUT_TYPE_P2SH;
+    } else if (type == OUTPUT_TYPE_STRING_BECH32) {
+        return OUTPUT_TYPE_BECH32;
+    } else {
+        return OUTPUT_TYPE_NONE;
+    }
+}
+
+
+const std::string& FormatOutputType(OutputType type)
+{
+    switch (type) {
+    case OUTPUT_TYPE_LEGACY: return OUTPUT_TYPE_STRING_LEGACY;
+    case OUTPUT_TYPE_P2SH: return OUTPUT_TYPE_STRING_P2SH;
+    case OUTPUT_TYPE_BECH32: return OUTPUT_TYPE_STRING_BECH32;
+    default: assert(false);
+    }
+}
+
+std::vector<CTxDestination> CWallet::GetAllDestinationsForKey(const CPubKey& key)
+{
+    CKeyID keyid = key.GetID();
+    if (key.IsCompressed()) {
+        CTxDestination segwit = WitnessV0KeyHash(keyid);
+        CTxDestination p2sh = CScriptID(GetScriptForDestination(segwit));
+        return std::vector<CTxDestination>{std::move(keyid), std::move(p2sh), std::move(segwit)};
+    } else {
+        return std::vector<CTxDestination>{std::move(keyid)};
+    }
+}
+
+CTxDestination CWallet::GetDestinationForKey(const CPubKey& key, OutputType type)
+{
+    switch (type) {
+    case OUTPUT_TYPE_LEGACY: return key.GetID();
+    case OUTPUT_TYPE_P2SH:
+    case OUTPUT_TYPE_BECH32: {
+        CTxDestination witdest = WitnessV0KeyHash(key.GetID());
+        CScript witprog = GetScriptForDestination(witdest);
+        // Check if the resulting program is solvable (i.e. doesn't use an uncompressed key)
+        if (!IsSolvable(*this, witprog)) return key.GetID();
+        // Even though CBasicKeyStore knows about P2WPKH redeemscripts implicitly, add it explicitly
+        // so that downgrades to software without this are supported (as long as the wallet file is
+        // up to date).
+        AddCScript(witprog);
```

Comment:
In commit "SegWit wallet support"

As brought up previously https://github.com/bitcoin/bitcoin/pull/11403#discussion_r145274394, this seems really ugly. Having GetDestinationForKey/GetDestinationForScript methods write to the wallet database when all you would expect them to do is get the address of a key or script doesn't seem safe or untuitive.

It seems like it would be better if the callers were responsible for adding CScripts. Since most callers are just calling GetDestinationForKey after reserving keys, I think if you added the CScripts in CWallet::ReserveKeyFromKeyPool it would take care of most of them.

If that isn't sufficien,t maybe there is some way to rename or wrap these methods so it is clear the calls are intended to actually update the database, rather than just compute addresses.

Reply:
I agree it's ugly, though I think it is temporary (it's not needing in my future design (see writeup), and it's also unnecessary once we drop support for downgrading to pre-implicit-SegWit wallet software - but that needs some thinking about versioning).

Splitting the choice for what to add and what destination to use up seems very inconvenient, though. The necessity of adding depends on the type of address.

I've renamed the functions to `Add...` rather than `Get...`, to reflect that they may change wallet state.